src/sql/workbench/services/connection/browser/connectionManagementService.ts

```
@@ -874,6 +876,61 @@ export class ConnectionManagementService extends Disposable implements IConnecti
+			this._logService.info(`Found pending reconnect promise for uri ${uri}, waiting.`);
+			try {
+				let previousConnectionResult = await previousReconnectPromise;
+				if (previousConnectionResult && previousConnectionResult.connected) {
+					this._logService.info(`Previous pending reconnection for uri ${uri} succeeded.`);
+					return true;
+				}
+				this._logService.info(`Previous pending reconnection for uri ${uri} failed with error ${previousConnectionResult.errorMessage}, error code: ${previousConnectionResult.errorCode}.`);
+			} catch (err) {
+				this._logService.info(`Previous pending reconnect promise for uri ${uri} is rejected with error ${err}, will attempt to reconnect if necessary.`);
+			}
+		}
+
+		let expiry = profile.options.expiresOn;
+		if (typeof expiry === 'number' && !Number.isNaN(expiry)) {
+			const currentTime = new Date().getTime() / 1000;
+			const maxTolerance = 2 * 60; // two minutes
+			if (expiry - currentTime < maxTolerance) {
+				this._logService.info(`Access token expired for connection ${profile.id} with uri ${uri}`);
+				try {
+					let connectionResultPromise = this.connect(profile, uri);
+					this._uriToReconnectPromiseMap[uri] = connectionResultPromise;
+					let connectionResult = await connectionResultPromise;
+					if (!connectionResult || !connectionResult.connected) {
+						this._logService.error(`Failed to refresh connection ${profile.id} with uri ${uri}, result: ${connectionResult}`);
+						throw new Error(`Connection Result: ${connectionResult}, error msg: ${connectionResult.errorMessage}, error code: ${connectionResult.errorCode}`);
+					}
+					this._logService.info(`Successfully refreshed token for connection ${profile.id} with uri ${uri}, result: ${connectionResult.connected} ${connectionResult.connectionProfile}, isConnected: ${this.isConnected(uri)}, ${this._connectionStatusManager.getConnectionProfile(uri)}`);
+					return true;
+				} catch (err) {
+					return Promise.reject(err);
```

You should be able to remove this catch and just let the exception bubble up

---

updated